src/simulator/tests/halo_simulator_test.py

```diff
@@ -113,6 +127,34 @@ def test_num_active_registers(self, cardinality, random_state, num_active):
             msg=f"The number of active registers for n={cardinality} is not correct.",
         )
 
+    @parameterized.named_parameters(
+        {
+            "testcase_name": "with_1_region",
+            "regions": {3: [1]},
+            "sample_size": 1,
+            "random_generator": np.random.default_rng(0),
+            "expected": {3: [1]},
+        },
+        {
+            "testcase_name": "with_5_regions_and_fake_rng",
+            "regions": {i: [i ** 2 + 1] for i in range(5)},
+            "sample_size": 20,
+            "random_generator": FakeRandomGenerator(),
+            "expected": {i: [n] for i, n in enumerate([1, 2, 5, 6, 6])},
```

Comment:
Why 1,2,5, 6, 6? 
IIU FakeRandomGenerator correctly, it'll be [0^2 + 1, 1^2 + 1, 2^2 + 1, 3^2 + 1, remaining] = [1, 2, 5, 10, 3]. Could you walk me through this?

Reply:
```python
class FakeRandomGenerator:
    def multivariate_hypergeometric(self, colors, nsample):
        samples = [0] * len(colors)
        index = 0

        while nsample:
            if samples[index] < colors[index]:
                samples[index] += 1
                nsample -= 1
            index = (index + 1) % len(samples)

        return samples
```

How the FakeRandomGenerator work is, it keeps going around the population list and draws one sample if the current slot has any items remained. 

Take the test case you quoted for example, the initial states are `colors = [1, 2, 5, 10, 17]`, `samples = [0, 0, 0, 0, 0]`, and `index = 0`. Inside the while-loop, `samples` changes as following:
    `[0, 0, 0, 0, 0]` -> `[1, 0, 0, 0, 0] (index = 0)` -> `[1, 1, 0, 0, 0] (index = 1)` -> `[1, 1, 1, 0, 0] (index = 2)` -> `[1, 1, 1, 1, 0] (index = 3)` -> `[1, 1, 1, 1, 1] (index = 4)`. Then `index` becomes 0 again. Since `samples[0]` is equal to `colors[0]`, i.e. no items available in slot-0, we do nothing and increment `index` to 1. Then,
    -> `[1, 2, 1, 1, 1]` -> ...  -> `[1, 2, 2, 2, 2]` -> `[1, 2, 3, 2, 2]` -> ...  -> `[1, 2, 5, 5, 5]` -> ... -> `[1, 2, 5, 6, 6]`. At this moment `nsample == 0`, exit the while-loop and return `samples`.

Hope that helps.

Reply:
I see I see. Sorry I mis-read the codes!
Thank you so much for the clear explanation! Let's leave it here for other reviewers.

Reply:
No problem at all! Glad to help 😃